### PR TITLE
[CAUTH-622] add two rules to be published when Adaptive MFA becomes GA

### PR DESCRIPTION
**Pull request by pmalouin**
### Description

Add two rules template presented when the administrator created a new rule to cover some specialized use cases for Adaptive MFA. While the out-of-the-box experience of Adaptive MFA does not strictly require creating custom rules, these two templates are useful for these two advanced cases:

- Adaptive MFA: this rule template covers an example of how to calculate a "custom" confidence score and trigger MFA based on that score instead of the default scoring scheme. This template is meant to be edited based on the administrator's requirements for scoring, but lays out how to access the granular confidence assessment results (in the example, the `ImpossibleTravel` assessment result).
- Require MFA Enrollment: this rule template covers the case when end-users that are not enrolled in MFA are invited to enroll in MFA on their next successful login attempt. The rule shows that we must trigger and enrollment conditionally based on the user's current MFA factors (if no current factors => enroll). This rule is also useful in different contexts than Adaptive MFA.

### Screenshots

image

##### Adaptive MFA
image

##### Require MFA Enrollment
image

### References

https://auth0team.atlassian.net/browse/CAUTH-622

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
>
> Please include any manual steps for testing end-to-end or functionality not covered by unit/integration tests.
>
> Also include details of the environment this PR was developed in (language/platform/browser version).

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

### CODE REVIEW

File: src/rules/adaptive-mfa.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * @title Adaptive MFA
+ * @overview Trigger multifactor authentication for a specific risk assessment result.
+ * @gallery true
+ * @category multifactor
+ *
+ * This rule is used to trigger multifactor authentication when a specific risk assessment result is detected.
+ *
+ * The `context.riskAssessment` attribute will be available only when the Adaptive MFA policy is enabled for your tenant. This feature is only available to customers that have subscribed to an Enterprise plan.
+ *
+ * See https://auth0.com/docs/<<<TODO>>> for more information.
```
Comment (pmalouin) [inline src/rules/adaptive-mfa.js:11]: FIXME: when links are available

File: src/rules/adaptive-mfa.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/**
+ * @title Adaptive MFA
+ * @overview Trigger multifactor authentication for a specific risk assessment result.
+ * @gallery true
+ * @category multifactor
+ *
+ * This rule is used to trigger multifactor authentication when a specific risk assessment result is detected.
+ *
+ * The `context.riskAssessment` attribute will be available only when the Adaptive MFA policy is enabled for your tenant. This feature is only available to customers that have subscribed to an Enterprise plan.
```
Comment (pmalouin) [inline src/rules/adaptive-mfa.js:9]: FIXME: plan name tbd

File: src/rules/mfa-require-enrollment.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * @title Require MFA Enrollment
+ * @overview Require MFA Enrollment on next login
+ * @gallery true
+ * @category multifactor
+ *
+ * This rule requires that any user not already enrolled in MFA will be presented with an enrollment prompt on their next login.
+ *
+ * This rule can be paired with the Adaptive MFA feature that prompts for an MFA challenge only when the login confidence is low. It can also be combined with another custom Rule that prompts for an MFA challenge under a specific condition.
+ *
+ * Note that the Adaptive MFA feature is only available to customers that have subscribed to an Enterprise plan.
```
Comment (pmalouin) [inline src/rules/mfa-require-enrollment.js:11]: FIXME: plan name tbd

File: rules.json [hunk 4]
```diff
@@ -164,6 +164,104 @@
       }
```
Comment (pmalouin) [inline rules.json:164]: For repo maintainer, should this generated file be part of the PR or not?

**Comment by randynasson** [comment]
@pmalouin The UI looks good.

File: src/rules/adaptive-mfa.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
+/**
+ * @title Adaptive MFA
```
Comment (hernanhht) [inline src/rules/adaptive-mfa.js:2]: I think we should use a more specific name like "Adaptive MFA - Impossible Travel Only"

Comment (pmalouin) [reply src/rules/adaptive-mfa.js:2]: @hernanhht 

So this template is meant to be an example of how to access the granular assessor results to determine a "custom" confidence score, and I've used Impossible Travel just to pick one out of the three we have. It's not like this is a use-case that we think customers will need/use, just a showcase of what's possible.

In that spirit, I think I prefer a more generic title, not specifically mentioning Impossible Travel, but to invite tenant admins to start from this example and work out their specific confidence calculation. Maybe we can be a more specific than just "Adaptive MFA". Ideas:

- Adaptive MFA with Custom Confidence (my preference)
- Adaptive MFA with Overridden Confidence
- Adaptive MFA with Special Confidence

wdyt?

Comment (hernanhht) [reply src/rules/adaptive-mfa.js:2]: Ok, "with custom confidence" is good enough

Comment (pmalouin) [reply src/rules/adaptive-mfa.js]: done

**Review by hernanhht — APPROVED** [review]

**Comment by joshcanhelp** [comment]
@pmalouin - When are you planning on this going out? Nothing too major here but if it's coming up soon, might be good to get an ESD in ahead of time to we're ready.

File: src/rules/adaptive-mfa-custom-confidence.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
+/**
+ * @title Adaptive MFA with Custom Confidence
+ * @overview Trigger multifactor authentication for a specific risk assessment result.
+ * @gallery true
+ * @category multifactor
+ *
+ * This rule is used to trigger multifactor authentication when a specific risk assessment result is detected.
+ *
+ * The `context.riskAssessment` attribute will be available only when the Adaptive MFA policy is enabled for your tenant. This feature is only available to customers that have subscribed to an Enterprise plan.
```
Comment (pmalouin) [inline src/rules/adaptive-mfa-custom-confidence.js:9]: FIXME: copy the sentence about Enterprise plan to a js doc inside the rule templates.

Comment (pmalouin) [reply src/rules/adaptive-mfa-custom-confidence.js:9]: And use the copy in this ticket as a comment: https://auth0team.atlassian.net/browse/CAUTH-611

File: rules.json (hunk 4 above)
Comment (joshcanhelp) [reply rules.json:164]: It can if it's going to be released on its own but it might cause conflicts if another release goes out before this one. Easy enough to pull in the latest and re-generate this file.

**Review by Deborah-Digges — APPROVED** [review]

File: rules.json [hunk 7]
```diff
@@ -772,4 +772,4 @@
...
     ]
   }
-]
\ No newline at end of file
+]
```
Comment (joshcanhelp) [inline rules.json:775]: Can you revert the change to this file?

File: src/rules/adaptive-mfa.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ * @overview Trigger multifactor authentication for a specific risk assessment result.
+ * @gallery true
+ * @category multifactor
+ *
+ * This rule is used to trigger multifactor authentication when a specific risk assessment result is detected.
+ *
+ * The `context.riskAssessment` attribute will be available only when Adaptive MFA is enabled for your tenant. Use of the Adaptive MFA feature requires an add-on for the Enterprise plan. Please contact sales with any questions.
```
Comment (joshcanhelp) [inline src/rules/adaptive-mfa.js:9]: **Nit:** URL/email to contact?

**Review by joshcanhelp — CHANGES_REQUESTED** [review]
Tiny change in `rules.json` but otherwise LGTM

**Review by joshcanhelp — APPROVED** [review]
Whitespace change should not block here.
